flink-ml-core/src/main/java/org/apache/flink/ml/common/datastream/DataStreamUtils.java

```diff
@@ -0,0 +1,244 @@
+     *
+     * @param <T> The class type of the input data stream.
+     * @param input The bounded input data stream.
+     * @return The result data stream that contains all the distinct values.
+     */
+    public static <T> DataStream<T> distinct(DataStream<T> input) {
+        return input.transform(
+                        "distinctInEachPartition",
+                        input.getType(),
+                        new DistinctPartitionOperator<>())
+                .setParallelism(input.getParallelism())
+                .transform(
+                        "distinctInFinalPartition",
+                        input.getType(),
+                        new DistinctPartitionOperator<>())
+                .setParallelism(1);
+    }
+
+    /**
+     * Applies a {@link MapPartitionFunction} on a bounded data stream.
+     *
+     * @param input The input data stream.
+     * @param func The user defined mapPartition function.
+     * @param <IN> The class type of the input element.
+     * @param <OUT> The class type of output element.
+     * @return The result data stream.
+     */
+    public static <IN, OUT> DataStream<OUT> mapPartition(
+            DataStream<IN> input, MapPartitionFunction<IN, OUT> func) {
+        TypeInformation<OUT> resultType =
+                TypeExtractor.getMapPartitionReturnTypes(func, input.getType(), null, true);
+        return input.transform("mapPartition", resultType, new MapPartitionOperator<>(func))
+                .setParallelism(input.getParallelism());
+    }
+
+    /**
+     * Sorts the elements in each partition of the input bounded data stream.
+     *
+     * @param input The input data stream.
+     * @param comparator The comparator used to sort the elements.
+     * @param <IN> The class type of input element.
+     * @return The sorted data stream.
+     */
+    public static <IN> DataStream<IN> sortPartition(
+            DataStream<IN> input, TypeComparator<IN> comparator) {
+        return input.transform(
+                        "sortPartition", input.getType(), new SortPartitionOperator<>(comparator))
+                .setParallelism(input.getParallelism());
+    }
+
+    /**
+     * A stream operator to compute the distinct values in each partition of the input bounded data
+     * stream.
+     */
+    static class DistinctPartitionOperator<T> extends AbstractStreamOperator<T>
+            implements OneInputStreamOperator<T, T>, BoundedOneInput {
+
+        private ListState<T> distinctLabelsState;
+
+        private Set<T> distinctLabels = new HashSet<>();
+
+        @Override
+        public void endInput() {
+            for (T distinctLabel : distinctLabels) {
+                output.collect(new StreamRecord<>(distinctLabel));
+            }
+            distinctLabelsState.clear();
+        }
+
+        @Override
+        public void processElement(StreamRecord<T> streamRecord) {
+            distinctLabels.add(streamRecord.getValue());
```

Comment:
I am not sure the performance of doing this is strictly better than using `distinctLabelsState.add(streamRecord.getValue())` directly. The reason is that `ListState` could choose to cache those data in memory and only flush data to disk when snapshot is invoked.

We can double check this with @gaoyunhaii.

Reply:
Using `distinctLabels` here is not simply for performance.

We actually need a updatable state --- We need to store all the distinct elements seen so far, but not ALL of the seen elements.

Reply:
I see. Sounds good.